content/posts/2019-11-07-deterministic-builds-with-clang-and-lld.md

```diff
@@ -0,0 +1,166 @@
+
+Visual Studio’s assemblers ml.exe and ml64.exe also insist on writing the current time into their output. In situations like this, where you can’t easily fix the tool to write the right output in the first place, you need to write wrappers that fix up the file after the fact. As an example, [ml.py](https://cs.chromium.org/chromium/src/build/toolchain/win/ml.py) is the wrapper the Chromium project uses to make ml’s output deterministic.
+
+macOS’s libtool and ld64 also insist on writing timestamps into their outputs. You can set the environment variable `ZERO_AR_DATE` to 1 in a wrapper to make their output deterministic, but that confuses lldb of older Xcode versions.
+
+Gcc sometimes uses random numbers in certain symbol mangling situations. Clang does not do this, so there’s no need to pass `-frandom-seed` to clang.
+
+It’s a good idea to make your build independent of environment variables as much as possible, so that accidental local changes in the environment don’t affect the build output. You should pass `/X` to clang-cl to make it ignore `%INCLUDE%` and explicitly pass system include directories via the `-imsvc` switch instead. Likewise, very new lld-link versions (LLVM 10 and newer, at the time of this writing still unreleased) understand the flag `/lldignoreenv` flag, which makes lld-link ignore the `%LIB%` environment variable; explicitly pass system library directories via `/libpath:`.
+
+## Footnote on embedding git hashes into the binary
+
+It might be tempting to embed the git commit hash or svn revision that a binary was built at into the binary’s `--version` output, or use the revision as a cache key to invalidate on-disk caches when the version changes.
+
+This doesn’t affect your build’s determinism, but it does affect the hit rate if you’re using deterministic builds to cache test run results. If your binary embeds the current commit, it is guaranteed to change on every single commit, and you won’t be able to cache test results across commits. Even commits that just fix typos in comments, add non-code documentation, or that only affect code used by some but not all of your binaries will change every binary.
+
+For cache invalidation, consider using something finer-grained, such as only the latest commit of the directory containing the cache handling code, or the hash of all source files containing the cache handling code.
+
+For `--version output`, if your build is fully deterministic, the hash of the binary itself (and its dynamic library dependencies) can serve as a stable version identifier. You can keep a map of binary hash to all commit hashes that produce that binary somewhere.
```

Comment:
This should be:

```
For `--version` output
```

(ie "output" should not be part of the code block)

Reply:
done.